File: src/main/java/com/github/cybellereaper/wizpets/core/service/PetServiceImpl.java
package com.github.cybellereaper.wizpets.core.service;

import com.github.cybellereaper.wizpets.api.ActivePet;
import com.github.cybellereaper.wizpets.api.DismissReason;
import com.github.cybellereaper.wizpets.api.PetRecord;
import com.github.cybellereaper.wizpets.api.StatSet;
import com.github.cybellereaper.wizpets.api.StatType;
import com.github.cybellereaper.wizpets.api.SummonReason;
import com.github.cybellereaper.wizpets.api.WizPetsApi;
import com.github.cybellereaper.wizpets.api.talent.PetTalent;
import com.github.cybellereaper.wizpets.api.talent.TalentFactory;
import com.github.cybellereaper.wizpets.api.talent.TalentRegistryView;
import com.github.cybellereaper.wizpets.api.timeline.PetLifecycleListener;
import com.github.cybellereaper.wizpets.core.config.PluginConfig;
import com.github.cybellereaper.wizpets.core.persistence.PetStorage;
import com.github.cybellereaper.wizpets.core.pet.ActivePetImpl;
import com.github.cybellereaper.wizpets.core.talent.TalentRegistryImpl;
import com.github.cybellereaper.wizpets.core.talent.defaults.ArcaneBurstTalent;
import com.github.cybellereaper.wizpets.core.talent.defaults.GuardianShellTalent;
import com.github.cybellereaper.wizpets.core.talent.defaults.HealingAuraTalent;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Random;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArraySet;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.ServicePriority;
import org.bukkit.plugin.java.JavaPlugin;

public final class PetServiceImpl implements WizPetsApi, Listener {
    private final JavaPlugin plugin;
    private final PluginConfig config;
    private final PetStorage storage;
    private final TalentRegistryImpl registry;
    private final Set<PetLifecycleListener> listeners = new CopyOnWriteArraySet<>();
    private final Map<UUID, ActivePetImpl> activePets = new ConcurrentHashMap<>();
    private final Random random = new Random(System.currentTimeMillis());

    public PetServiceImpl(JavaPlugin plugin) {
        this.plugin = Objects.requireNonNull(plugin, "plugin");
        this.config = new PluginConfig(plugin.getConfig());
        this.storage = new PetStorage(plugin);
        this.registry = new TalentRegistryImpl();
        registerDefaults();
        plugin.getServer().getServicesManager().register(WizPetsApi.class, this, plugin, ServicePriority.Normal);
    }

    public JavaPlugin getPlugin() {
        return plugin;
    }

    public PluginConfig getConfigValues() {
        return config;
    }

    @Override
    public ActivePet activePet(Player player) {
        return activePets.get(player.getUniqueId());
    }

    @Override
    public PetRecord storedPet(Player player) {
        return storage.load(player);
    }

    @Override
    public ActivePet summon(Player player, SummonReason reason) {
        ActivePetImpl existing = activePets.remove(player.getUniqueId());
        if (existing != null) {
            storage.save(player, existing.toRecord());
            existing.remove(true);
        }
        PetRecord loaded = storage.load(player);
        if (loaded == null) {
            loaded = createNewRecord(player);
        }
        ResolvedTalents resolved = resolveTalents(loaded);
        ActivePetImpl pet = new ActivePetImpl(this, player, resolved.record(), resolved.talents());
        pet.spawn();
        activePets.put(player.getUniqueId(), pet);
        storage.save(player, pet.toRecord());
        listeners.forEach(listener -> listener.onSummoned(player, pet, reason));
        plugin.getLogger().fine(() -> "Summoned pet for " + player.getName() + " via " + reason);
        return pet;
    }

    @Override
    public boolean dismiss(Player player, DismissReason reason) {
        return dismiss(player, reason, reason == DismissReason.PLUGIN_DISABLE);
    }

    public void dismiss(Player player, boolean internal) {
        dismiss(player, internal ? DismissReason.PLAYER_QUIT : DismissReason.MANUAL, internal);
    }

    private boolean dismiss(Player player, DismissReason reason, boolean persistFlight) {
        ActivePetImpl pet = activePets.remove(player.getUniqueId());
        if (pet == null) {
            return false;
        }
        PetRecord record = pet.toRecord();
        storage.save(player, record);
        listeners.forEach(listener -> listener.onDismissed(player, record, reason));
        pet.remove(persistFlight);
        plugin.getLogger().fine(() -> "Dismissed pet for " + player.getName() + " via " + reason);
        return true;
    }

    @Override
    public void persist(Player player) {
        ActivePetImpl pet = activePets.get(player.getUniqueId());
        if (pet == null) {
            return;
        }
        PetRecord record = pet.toRecord();
        storage.save(player, record);
        listeners.forEach(listener -> listener.onPersist(player, record));
    }

    @Override
    public TalentRegistryView talents() {
        return registry;
    }

    @Override
    public void registerTalent(TalentFactory factory, boolean replace) {
        registry.register(factory, replace);
        refreshActivePets();
    }

    @Override
    public void unregisterTalent(String id) {
        registry.unregister(id);
        refreshActivePets();
    }

    @Override
    public void addListener(PetLifecycleListener listener) {
        listeners.add(listener);
    }

    @Override
    public void removeListener(PetLifecycleListener listener) {
        listeners.remove(listener);
    }

    public boolean mount(Player player) {
        ActivePetImpl pet = activePets.get(player.getUniqueId());
        if (pet == null) {
            return false;
        }
        boolean mounted = pet.mount();
        if (mounted) {
            persist(player);
        }
        return mounted;
    }

    public boolean dismount(Player player) {
        ActivePetImpl pet = activePets.get(player.getUniqueId());
        if (pet == null) {
            return false;
        }
        boolean result = pet.dismount();
        if (result) {
            persist(player);
        }
        return result;
    }

    public boolean enableFlight(Player player) {
        ActivePetImpl pet = activePets.get(player.getUniqueId());
        if (pet == null) {
            return false;
        }
        boolean started = pet.startFlying();
        if (started) {
            persist(player);
        }
        return started;
    }

    public boolean disableFlight(Player player) {
        ActivePetImpl pet = activePets.get(player.getUniqueId());
        if (pet == null) {
            return false;
        }
        boolean stopped = pet.stopFlying();
        if (stopped) {
            persist(player);
        }
        return stopped;
    }

    public void restoreOnlinePlayers() {
        Bukkit.getOnlinePlayers().forEach(player -> {
            if (config.isAutoSummon()) {
                summon(player, SummonReason.RESTORE);
            }
        });
    }

    public void shutdown() {
        List<ActivePetImpl> snapshot = new ArrayList<>(activePets.values());
        snapshot.forEach(pet -> {
            Player owner = pet.getOwner();
            storage.save(owner, pet.toRecord());
            pet.remove(false);
            listeners.forEach(listener -> listener.onDismissed(owner, pet.toRecord(), DismissReason.PLUGIN_DISABLE));
        });
        activePets.clear();
    }

    public void unregister() {
        plugin.getServer().getServicesManager().unregister(WizPetsApi.class, this);
    }

    public void breed(Player player, Player partner) {
        PetRecord playerRecord = storage.load(player);
        if (playerRecord == null) {
            playerRecord = createNewRecord(player);
        }
        PetRecord partnerRecord = storage.load(partner);
        if (partnerRecord == null) {
            return;
        }

        int generation = Math.max(playerRecord.getGeneration(), partnerRecord.getGeneration()) + 1;
        PetRecord childRecord = new PetRecord(
            MessageFormat.format("{0}'s Hatchling", player.getName()),
            playerRecord.getEvs().breedWith(partnerRecord.getEvs(), random),
            playerRecord.getIvs().breedWith(partnerRecord.getIvs(), random),
            registry.inherit(playerRecord.getTalentIds(), partnerRecord.getTalentIds(), random),
            generation,
            playerRecord.getBreedCount() + 1,
            playerRecord.isMountUnlocked() || partnerRecord.isMountUnlocked(),
            playerRecord.isFlightUnlocked() || partnerRecord.isFlightUnlocked()
        );

        storage.save(player, childRecord);
        listeners.forEach(listener -> listener.onPersist(player, childRecord));
        summon(player, SummonReason.BREEDING_REFRESH);

        PetRecord updatedPartner = partnerRecord.withBreedCount(partnerRecord.getBreedCount() + 1);
        storage.save(partner, updatedPartner);
        listeners.forEach(listener -> listener.onPersist(partner, updatedPartner));
        ActivePetImpl partnerPet = activePets.get(partner.getUniqueId());
        if (partnerPet != null) {
            ResolvedTalents resolved = resolveTalents(updatedPartner);
            storage.save(partner, resolved.record());
            partnerPet.update(resolved.record(), resolved.talents());
        }
    }

    public List<String> debugLines(Player player) {
        PetRecord record = storage.load(player);
        if (record == null) {
            return List.of("§cNo stored pet data found.");
        }
        String stats = buildStatsSummary(record);
        return List.of(
            "§7==== §dPet Debug §7====",
            "§7Name: §f" + record.getDisplayName(),
            "§7Generation: §f" + record.getGeneration() + "  §7Breeds: §f" + record.getBreedCount(),
            "§7Talents: §f" + (record.getTalentIds().isEmpty() ? "None" : String.join(", ", record.getTalentIds())),
            "§7Mount unlocked: §f" + record.isMountUnlocked(),
            "§7Flight unlocked: §f" + record.isFlightUnlocked(),
            "§7Base Stats: §f" + stats
        );
    }

    private String buildStatsSummary(PetRecord record) {
        StringBuilder builder = new StringBuilder();
        StatType[] types = StatType.values();
        for (int i = 0; i < types.length; i++) {
            StatType type = types[i];
            double value = record.getEvs().value(type) / 4.0 + record.getIvs().value(type);
            builder.append(type.getDisplayName())
                .append(':')
                .append(' ')
                .append(String.format(Locale.US, "%.2f", value));
            if (i < types.length - 1) {
                builder.append(", ");
            }
        }
        return builder.toString();
    }

    private PetRecord createNewRecord(Player player) {
        PetRecord record = new PetRecord(
            config.getDefaultPetName().replace("{player}", player.getName()),
            StatSet.randomEV(random),
            StatSet.randomIV(random),
            registry.roll(random),
            1,
            0,
            false,
            false
        );
        storage.save(player, record);
        return record;
    }

    private void refreshActivePets() {
        activePets.values().forEach(pet -> {
            Player owner = pet.getOwner();
            PetRecord record = storage.load(owner);
            if (record == null) {
                record = pet.toRecord();
            }
            ResolvedTalents resolved = resolveTalents(record);
            storage.save(owner, resolved.record());
            pet.update(resolved.record(), resolved.talents());
        });
    }

    private void registerDefaults() {
        registry.register(HealingAuraTalent::new);
        registry.register(GuardianShellTalent::new);
        registry.register(ArcaneBurstTalent::new);
    }

    private ResolvedTalents resolveTalents(PetRecord record) {
        PetRecord currentRecord = record;
        List<PetTalent> talents = registry.instantiate(currentRecord.getTalentIds());
        if (talents.size() != currentRecord.getTalentIds().size() || talents.isEmpty()) {
            List<String> ids = registry.roll(random);
            currentRecord = currentRecord.withTalentIds(ids);
            talents = registry.instantiate(ids);
        }
        return new ResolvedTalents(currentRecord, talents);
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        if (config.isAutoSummon()) {
            Bukkit.getScheduler().runTask(plugin, () -> summon(event.getPlayer(), SummonReason.AUTO_SUMMON));
        }
    }

    @EventHandler
    public void onPlayerQuit(PlayerQuitEvent event) {
        dismiss(event.getPlayer(), DismissReason.PLAYER_QUIT, false);
    }

    private record ResolvedTalents(PetRecord record, List<PetTalent> talents) {
    }
}

File: src/main/kotlin/com/github/cybellereaper/wizpets/core/service/PetServiceImpl.kt
package com.github.cybellereaper.wizpets.core.service

import com.github.cybellereaper.wizpets.api.ActivePet
import com.github.cybellereaper.wizpets.api.DismissReason
import com.github.cybellereaper.wizpets.api.PetRecord
import com.github.cybellereaper.wizpets.api.StatSet
import com.github.cybellereaper.wizpets.api.StatType
import com.github.cybellereaper.wizpets.api.SummonReason
import com.github.cybellereaper.wizpets.api.WizPetsApi
import com.github.cybellereaper.wizpets.api.talent.PetTalent
import com.github.cybellereaper.wizpets.api.talent.TalentFactory
import com.github.cybellereaper.wizpets.api.talent.TalentRegistryView
import com.github.cybellereaper.wizpets.api.timeline.PetLifecycleListener
import com.github.cybellereaper.wizpets.core.config.PluginConfig
import com.github.cybellereaper.wizpets.core.persistence.PetStorage
import com.github.cybellereaper.wizpets.core.pet.ActivePetImpl
import com.github.cybellereaper.wizpets.core.talent.TalentRegistryImpl
import com.github.cybellereaper.wizpets.core.talent.defaults.ArcaneBurstTalent
import com.github.cybellereaper.wizpets.core.talent.defaults.GuardianShellTalent
import com.github.cybellereaper.wizpets.core.talent.defaults.HealingAuraTalent
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.plugin.ServicePriority
import org.bukkit.plugin.java.JavaPlugin
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max
import kotlin.random.Random

class PetServiceImpl(val plugin: JavaPlugin) : WizPetsApi, Listener {
    val config = PluginConfig(plugin.config)
    private val storage = PetStorage(plugin)
    private val registry = TalentRegistryImpl()
    private val listeners = mutableSetOf<PetLifecycleListener>()
    private val activePets = ConcurrentHashMap<UUID, ActivePetImpl>()
    private val random = Random(System.currentTimeMillis())

    init {
        registerDefaults()
        plugin.server.servicesManager.register(WizPetsApi::class.java, this, plugin, ServicePriority.Normal)
    }

    override fun activePet(player: Player): ActivePet? = activePets[player.uniqueId]

    override fun storedPet(player: Player): PetRecord? = storage.load(player)

    override fun summon(player: Player, reason: SummonReason): ActivePet {
        val existing = activePets.remove(player.uniqueId)
        if (existing != null) {
            storage.save(player, existing.toRecord())
            existing.remove(persistFlight = true)
        }
        val loaded = storage.load(player) ?: createNewRecord(player)
        val resolved = resolveTalents(loaded)
        val pet = ActivePetImpl(this, player, resolved.first, resolved.second)
        pet.spawn()
        activePets[player.uniqueId] = pet
        storage.save(player, pet.toRecord())
        listeners.forEach { it.onSummoned(player, pet, reason) }
        plugin.logger.fine("Summoned pet for ${player.name} via $reason")
        return pet
    }

    override fun dismiss(player: Player, reason: DismissReason): Boolean = dismissInternal(player, reason, persistFlight = reason == DismissReason.PLUGIN_DISABLE)

    internal fun dismiss(player: Player, internal: Boolean = false) {
        dismissInternal(player, if (internal) DismissReason.PLAYER_QUIT else DismissReason.MANUAL, persistFlight = internal)
    }

    private fun dismissInternal(player: Player, reason: DismissReason, persistFlight: Boolean): Boolean {
        val pet = activePets.remove(player.uniqueId) ?: return false
        val record = pet.toRecord()
        storage.save(player, record)
        listeners.forEach { it.onDismissed(player, record, reason) }
        pet.remove(persistFlight)
        plugin.logger.fine("Dismissed pet for ${player.name} via $reason")
        return true
    }

    override fun persist(player: Player) {
        val pet = activePets[player.uniqueId] ?: return
        val record = pet.toRecord()
        storage.save(player, record)
        listeners.forEach { it.onPersist(player, record) }
    }

    override fun talents(): TalentRegistryView = registry

    override fun registerTalent(factory: TalentFactory, replace: Boolean) {
        registry.register(factory, replace)
        refreshActivePets()
    }

    override fun unregisterTalent(id: String) {
        registry.unregister(id)
        refreshActivePets()
    }

    override fun addListener(listener: PetLifecycleListener) {
        listeners += listener
    }

    override fun removeListener(listener: PetLifecycleListener) {
        listeners -= listener
    }

    fun mount(player: Player): Boolean {
        val pet = activePets[player.uniqueId] ?: return false
        val mounted = pet.mount()
        if (mounted) {
            persist(player)
        }
        return mounted
    }

    fun dismount(player: Player): Boolean {
        val pet = activePets[player.uniqueId] ?: return false
        val result = pet.dismount()
        if (result) {
            persist(player)
        }
        return result
    }

    fun enableFlight(player: Player): Boolean {
        val pet = activePets[player.uniqueId] ?: return false
        val started = pet.startFlying()
        if (started) {
            persist(player)
        }
        return started
    }

    fun disableFlight(player: Player): Boolean {
        val pet = activePets[player.uniqueId] ?: return false
        val stopped = pet.stopFlying()
        if (stopped) {
            persist(player)
        }
        return stopped
    }

    fun restoreOnlinePlayers() {
        Bukkit.getOnlinePlayers().forEach { player ->
            if (config.autoSummon) {
                summon(player, SummonReason.RESTORE)
            }
        }
    }

    fun shutdown() {
        activePets.values.toList().forEach { pet ->
            val owner = pet.owner
            storage.save(owner, pet.toRecord())
            pet.remove(persistFlight = false)
            listeners.forEach { it.onDismissed(owner, pet.toRecord(), DismissReason.PLUGIN_DISABLE) }
        }
        activePets.clear()
    }

    fun unregister() {
        plugin.server.servicesManager.unregister(WizPetsApi::class.java, this)
    }

    fun breed(player: Player, partner: Player) {
        val playerRecord = storage.load(player) ?: createNewRecord(player)
        val partnerRecord = storage.load(partner) ?: return

        val generation = max(playerRecord.generation, partnerRecord.generation) + 1
        val childRecord = PetRecord(
            displayName = "${player.name}'s Hatchling",
            evs = playerRecord.evs.breedWith(partnerRecord.evs, random),
            ivs = playerRecord.ivs.breedWith(partnerRecord.ivs, random),
            talentIds = registry.inherit(playerRecord.talentIds, partnerRecord.talentIds, random),
            generation = generation,
            breedCount = playerRecord.breedCount + 1,
            mountUnlocked = playerRecord.mountUnlocked || partnerRecord.mountUnlocked,
            flightUnlocked = playerRecord.flightUnlocked || partnerRecord.flightUnlocked
        )

        storage.save(player, childRecord)
        listeners.forEach { it.onPersist(player, childRecord) }
        summon(player, SummonReason.BREEDING_REFRESH)

        val updatedPartner = partnerRecord.copy(breedCount = partnerRecord.breedCount + 1)
        storage.save(partner, updatedPartner)
        listeners.forEach { it.onPersist(partner, updatedPartner) }
        activePets[partner.uniqueId]?.update(updatedPartner, resolveTalents(updatedPartner).second)
    }

    fun debugLines(player: Player): List<String> {
        val record = storage.load(player) ?: return listOf("§cNo stored pet data found.")
        val stats = StatType.entries.joinToString(", ") { "${it.displayName}: ${"%.2f".format(record.evs[it] / 4.0 + record.ivs[it])}" }
        return listOf(
            "§7==== §dPet Debug §7====",
            "§7Name: §f${record.displayName}",
            "§7Generation: §f${record.generation}  §7Breeds: §f${record.breedCount}",
            "§7Talents: §f${if (record.talentIds.isEmpty()) "None" else record.talentIds.joinToString(", ")}",
            "§7Mount unlocked: §f${record.mountUnlocked}",
            "§7Flight unlocked: §f${record.flightUnlocked}",
            "§7Base Stats: §f$stats"
        )
    }

    private fun createNewRecord(player: Player): PetRecord {
        val record = PetRecord(
            displayName = config.defaultPetName.replace("{player}", player.name),
            evs = StatSet.randomEV(random),
            ivs = StatSet.randomIV(random),
            talentIds = registry.roll(random),
            generation = 1,
            breedCount = 0,
            mountUnlocked = false,
            flightUnlocked = false
        )
        storage.save(player, record)
        return record
    }

    private fun refreshActivePets() {
        activePets.values.forEach { pet ->
            val owner = pet.owner
            val record = storage.load(owner) ?: pet.toRecord()
            val resolved = resolveTalents(record)
            storage.save(owner, resolved.first)
            pet.update(resolved.first, resolved.second)
        }
    }

    private fun registerDefaults() {
        registry.register({ HealingAuraTalent() })
        registry.register({ GuardianShellTalent() })
        registry.register({ ArcaneBurstTalent() })
    }

    private fun resolveTalents(record: PetRecord): Pair<PetRecord, List<PetTalent>> {
        var currentRecord = record
        var talents = registry.instantiate(currentRecord.talentIds)
        if (talents.size != currentRecord.talentIds.size || talents.isEmpty()) {
            val ids = registry.roll(random)
            currentRecord = currentRecord.copy(talentIds = ids)
            talents = registry.instantiate(ids)
        }
        return currentRecord to talents
    }

    @EventHandler
    fun onPlayerJoin(event: PlayerJoinEvent) {
        if (config.autoSummon) {
            Bukkit.getScheduler().runTask(plugin, Runnable { summon(event.player, SummonReason.AUTO_SUMMON) })
        }
    }

    @EventHandler
    fun onPlayerQuit(event: PlayerQuitEvent) {
        dismissInternal(event.player, DismissReason.PLAYER_QUIT, persistFlight = false)
    }
}
